test/gc.js

```diff
@@ -0,0 +1,101 @@
+'use strict'
+/* global WeakRef, FinalizationRegistry */
+
+const { test } = require('tap')
+const { createServer } = require('net')
+const { Client, Pool } = require('..')
+const { kClients } = require('../lib/core/symbols')
+
+const SKIP = typeof WeakRef === 'undefined' || typeof FinalizationRegistry === 'undefined'
+
+test('gc should collect the client if, and only if, there are no active sockets', { skip: SKIP }, t => {
+  t.plan(4)
+
+  const server = createServer((socket) => {
+    socket.write('HTTP/1.1 200 OK\r\n')
+    socket.write('Content-Length: 0\r\n')
+    socket.write('Keep-Alive: timeout=1s\r\n')
+    socket.write('Connection: keep-alive\r\n')
+    socket.write('\r\n\r\n')
+  })
+  t.teardown(server.close.bind(server))
+
+  let weakRef
+  let disconnected = false
+
+  const registry = new FinalizationRegistry((data) => {
+    t.strictEqual(data, 'test')
+    t.strictEqual(disconnected, true)
+    t.strictEqual(weakRef.deref(), undefined)
+  })
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`, {
+      keepAliveTimeoutThreshold: 500
+    })
+    client.once('disconnect', () => {
+      disconnected = true
+      setImmediate(() => {
+        global.gc()
+      })
+    })
+
+    weakRef = new WeakRef(client)
+    registry.register(client, 'test')
+
+    client.request({
+      path: '/',
+      method: 'GET'
+    }, (err, { body }) => {
+      t.error(err)
+      body.resume()
+    })
+  })
+})
+
+test('gc should collect the pool if, and only if, there are no active sockets', { skip: SKIP }, t => {
+  t.plan(4)
+
+  const server = createServer((socket) => {
+    socket.write('HTTP/1.1 200 OK\r\n')
+    socket.write('Content-Length: 0\r\n')
+    socket.write('Keep-Alive: timeout=1s\r\n')
+    socket.write('Connection: keep-alive\r\n')
+    socket.write('\r\n\r\n')
+  })
+  t.teardown(server.close.bind(server))
+
+  let weakRef
+  let disconnected = false
+
+  const registry = new FinalizationRegistry((data) => {
+    t.strictEqual(data, 'test')
+    t.strictEqual(disconnected, true)
+    t.strictEqual(weakRef.deref(), undefined)
+  })
+
+  server.listen(0, () => {
+    const pool = new Pool(`http://localhost:${server.address().port}`, {
+      connections: 1,
+      keepAliveTimeoutThreshold: 500
+    })
+
+//    pool.once('disconnect', () => {
```

Comment:
To test the same way we did for the `Client`, maybe the `Pool` should emit a `disconnect` event when all clients have disconnected, wdyt?

Reply:
I think pool should emit connect/disconnect with the client as first argument, i.e. simply forwarding the client events.

Reply:
Ok. I think it would be better to open a separate issue/PR for this and then rebase the work we are doing here.